src/commands/start.command.ts

```diff
@@ -115,15 +110,99 @@ export const start = async (apps: AppDefinition[]) => {
   });
 
   process.on("SIGINT", function () {
-    console.log("shutting down")
-    server.close();
+    shutdown();
+  });
+
+  process.on("SIGTERM", function () {
+    shutdown();
+  });
+
+  const events = ["beforeExit", "disconnected", "message", "rejectionHandled", "uncaughtException", "exit", "SIGABRT", "SIGHUP", "SIGPWR", "SIGQUIT"];
+
+  events.forEach((eventName) => {
+    process.on(eventName, (...args) => {
+      console.log('Unhandled error event ' + eventName + ' was called with args : ' + args.join(','));
+      shutdown();
+    });
+  });
+
+
+  const shutdown = async (s?: Socket) => {
+    console.log("shutting down");
     shuttingDown = true;
-    for (const name in processes) {
-      const p = processes[name];
-      if (!p.killed) {
-        console.log("killing " + name);
-        p.kill("SIGINT");
+    await Promise.all(Object.values(processes).map(async p => {
+      if (p.pid) {
+        const list: { [key: string]: string[] } = await getChildProcesses(p.pid.toString(), { [p.pid]: [] }, { [p.pid]: 1 });
+        killProcesses(list);
       }
+    }));
+    server.close();
+    s?.destroy();
+    process.exit();
+  }
+
+  const killProcesses = (tree: { [key: string]: string[] }) => {
+    const killed: { [key: string]: number } = {};
+    Object.keys(tree).map((pid) => {
+      tree[pid].map((childPid) => {
+        if (!killed[childPid]) {
+          killPid(childPid);
+          killed[childPid] = 1;
+        }
+        if (!killed[pid]) {
+          killPid(pid);
+          killed[pid] = 1;
+        }
+      })
+    })
+    return;
+  }
+
+  const killPid = (pid: string) => {
+    try {
+      process.kill(parseInt(pid, 10), "SIGINT");
+    } catch (err) {
+      console.error(err);
     }
-  });
+  }
+
+  const getChildProcesses = async (parentPid: string, tree: { [key: string]: string[] }, pidsToProcess: { [key: string]: number }): Promise<{ [key: string]: string[] }> => {
+    return new Promise((resolve, reject) => {
+      const ps = spawn('pgrep', ['-P', parentPid]);
```

Comment:
Using a system command and parsing it's output is fragile (cross platform, cross different versions)

isn't there a better way to do that?

Reply:
I tried the most popular packages which build process trees but they all failed to include some grandchild-processes
But you are right the cross platform issue needs to be handled in any case

Reply:
perhaps this technique would be a solution: https://azimi.me/2014/12/31/kill-child_process-node-js.html

Reply:
Strange...
I tried that approach but it didn't work. Now it does. This simplifies the implementation a lot!